daemon/api_users.go

```diff
@@ -239,44 +230,24 @@ func postUsers(c *Command, r *http.Request, user *auth.UserState) Response {
 }
 
 func removeUser(c *Command, username string, opts postUserDeleteData) Response {
-	// TODO: allow to remove user entries by email as well
-
-	// catch silly errors
-	if username == "" {
-		return BadRequest("need a username to remove")
-	}
-	// check the user is known to snapd
 	st := c.d.overlord.State()
 	st.Lock()
-	_, err := auth.UserByUsername(st, username)
-	st.Unlock()
-	if err == auth.ErrInvalidUser {
-		return BadRequest("user %q is not known", username)
-	}
-	if err != nil {
-		return InternalError(err.Error())
-	}
+	defer st.Unlock()
 
-	// first remove the system user
-	if err := osutilDelUser(username, &osutil.DelUserOptions{ExtraUsers: !release.OnClassic}); err != nil {
-		return InternalError(err.Error())
-	}
-
-	// then the UserState
-	st.Lock()
-	u, err := auth.RemoveUserByUsername(st, username)
-	st.Unlock()
-	// ErrInvalidUser means "not found" in this case
-	if err != nil && err != auth.ErrInvalidUser {
-		return InternalError(err.Error())
-	}
-
-	result := map[string]interface{}{
-		"removed": []userResponseData{
-			{ID: u.ID, Username: u.Username, Email: u.Email},
-		},
+	if u, err := deviceStateRemoveUser(st, username); err != nil {
+		if _, ok := err.(*devicestate.UserError); ok {
+			return BadRequest(err.Error())
+		} else {
+			return InternalError(err.Error())
+		}
+	} else {
```

Comment:
nitpick: usually we wouldn't have a else here

Reply:
Ah nice, I simplified several unneeded `else` conditions